src/network/uri/uri.hpp

```diff
@@ -215,68 +215,125 @@ class uri {
   const_iterator end() const;
 
   /**
+   * \brief Tests whether this URI has a scheme component.
+   * \return \c true if the URI has a scheme, \c false otherwise.
+   */
+  bool has_scheme() const;
+
+  /**
    * \brief Returns the URI scheme.
-   * \return The scheme, if it exists, or nullopt.
+   * \return The scheme.
+   * \pre has_scheme()
+   */
+  string_view scheme() const;
+
+  /**
+   * \brief Tests whether this URI has a user info component.
+   * \return \c true if the URI has a user info, \c false otherwise.
    */
-  optional<string_view> scheme() const;
+  bool has_user_info() const;
 
   /**
    * \brief Returns the URI user info.
-   * \return The user info, if it exists, or nullopt.
+   * \return The user info.
+   * \pre has_user_info()
    */
-  optional<string_view> user_info() const;
+  string_view user_info() const;
+
+  /**
+   * \brief Tests whether this URI has a host component.
+   * \return \c true if the URI has a host, \c false otherwise.
+   */
+  bool has_host() const;
 
   /**
    * \brief Returns the URI host.
-   * \return The host, if it exists, or nullopt.
+   * \return The host.
+   * \pre has_host()
    */
-  optional<string_view> host() const;
+  string_view host() const;
+
+  /**
+   * \brief Tests whether this URI has a port component.
+   * \return \c true if the URI has a port, \c false otherwise.
+   */
+  bool has_port() const;
 
   /**
    * \brief Returns the URI port.
-   * \return The port, if it exists, or nullopt.
+   * \return The port.
+   * \pre has_port()
    */
-  optional<string_view> port() const;
+  string_view port() const;
 
   /**
    * \brief Returns the URI port as an integer.
-   * \return The port, if it exists, or nullopt.
+   * \return The port number.
+   * \pre has_port()
    */
   template <typename IntT>
-  optional<IntT> port(typename std::is_integral<IntT>::type * = 0) const {
-    if (auto p = port()) {
-      try {
-        return std::stoi(string_type(std::begin(*p), std::end(*p)));
-      } catch (std::exception &) {
-        return optional<IntT>();
-      }
+  IntT port(typename std::is_integral<IntT>::type * = 0) const {
```

Comment:
Re: your comment the port accessor:

Remember that this already an overload. "string_view port() const" also exists. I provided another accessor that does the conversion to an integer type. This second function might either return an optional, or throw an exception.